File: shared-business-infrastructure/warehousing/warehousing/shared-libraries/Backend/Java/core-libraries/shared-exceptions/src/main/java/com/gogidix/shared/exceptions/domain/exception/BaseException.java
package com.gogidix.shared.exceptions.domain.exception;

import lombok.Getter;
import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.Map;

/**
 * Base exception for all Gogidix exceptions
 * Provides common exception handling functionality with rich context
 */
@Getter
public abstract class BaseException extends RuntimeException {
    
    private final String errorCode;
    private final int httpStatus;
    private final LocalDateTime timestamp;
    private final Map<String, Object> context;
    private final String category;
    
    protected BaseException(String message, String errorCode, int httpStatus, String category) {
        super(message);
        this.errorCode = errorCode;
        this.httpStatus = httpStatus;
        this.category = category;
        this.timestamp = LocalDateTime.now();
        this.context = new HashMap<>();
    }
    
    protected BaseException(String message, Throwable cause, String errorCode, int httpStatus) {
        super(message, cause);
        this.errorCode = errorCode;
        this.httpStatus = httpStatus;
        this.category = "GENERAL";
        this.timestamp = LocalDateTime.now();
        this.context = new HashMap<>();
    }
    
    protected BaseException(String message, Throwable cause, String errorCode, int httpStatus, String category) {
        super(message, cause);
        this.errorCode = errorCode;
        this.httpStatus = httpStatus;
        this.category = category;
        this.timestamp = LocalDateTime.now();
        this.context = new HashMap<>();
    }
    
    public void addContext(String key, Object value) {
        this.context.put(key, value);
    }
    
    public boolean isCritical() {
        return httpStatus >= 500;
    }
    
    public boolean isRetryable() {
        return httpStatus == 503 || httpStatus == 504;
    }
}

File: management-domain/Finance-department/Backend/Nodes/finance-automation-service/invoice-processing-service/src/shared/exceptions/base.exception.ts
import { HttpException, HttpStatus } from '@nestjs/common';

/**
 * Base exception class for all custom exceptions
 */
export abstract class BaseException extends HttpException {
  readonly errorCode: string;
  readonly context?: Record<string, unknown>;

  constructor(
    message: string,
    statusCode: HttpStatus,
    errorCode: string,
    context?: Record<string, unknown>,
  ) {
    super(
      {
        statusCode,
        message,
        errorCode,
        context,
        timestamp: new Date().toISOString(),
      },
      statusCode,
    );
    this.errorCode = errorCode;
    this.context = context;
    this.name = this.constructor.name;
  }
}
